### PR TITLE
Remove superfluous use of std::vector for std::future

### LINKED ISSUE #16888: [FEA] Remove superfluous use of std::vector for std::future
**Is your feature request related to a problem? Please describe.**
`std::vector<std::future>` [used here](std::vector<std::future<size_t>>) is unnecessary since the vector only contains a single element. The use of vector should be removed.

**Describe the solution you'd like**
Remove the vector.

**Describe alternatives you've considered**
N/A.

**Additional context**
N/A.

### PR DESCRIPTION
**Pull request by kingcrimsontianyu**
## Description
This PR addresses #16888 , where a superfluous use of `std::vector` should be removed.

closes #16888

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/parquet/reader_impl_preprocess.cu [hunk 1]
```diff
@@ -276,8 +276,14 @@ void generate_depth_remappings(
...
-    for (auto& task : read_tasks) {
-      task.wait();
+    std::vector<std::thread> threads;
+    for (std::size_t task_idx = 0; task_idx < read_tasks.size(); ++task_idx) {
+      threads.emplace_back([](std::future<size_t>& task) { task.wait(); },
+                           std::ref(read_tasks[task_idx]));
+    }
+
+    for (auto&& thread : threads) {
+      thread.join();
```
[the remark this answers is not in the document]
Comment (kingcrimsontianyu) [reply cpp/src/io/parquet/reader_impl_preprocess.cu:286]: Futures with a deferred launch policy conduct lazy evaluation such that the task is not run until `wait()` is called. So assuming full concurrency can be achieved, if we have `N` tasks, each taking `T` to finish, calling `future.wait()` in sequence would take `NT` to complete, whereas using threads would take `T`.
PS: Full concurrency can't be achieved for this problem, so I have observed no performance improvement by switching to using threads. But I just want to point out a potential issue with calling `future.wait()` in sequence.

File: cpp/src/io/parquet/reader_impl_preprocess.cu [hunk 2]
```diff
@@ -276,8 +276,14 @@ void generate_depth_remappings(
     }
   }
   auto sync_fn = [](decltype(read_tasks) read_tasks) {
-    for (auto& task : read_tasks) {
-      task.wait();
+    std::vector<std::thread> threads;
+    for (std::size_t task_idx = 0; task_idx < read_tasks.size(); ++task_idx) {
+      threads.emplace_back([](std::future<size_t>& task) { task.wait(); },
```
Comment (vuule) [inline cpp/src/io/parquet/reader_impl_preprocess.cu:281]: hm, I don't think we want this. We're creating a thread for each read task. Maybe using a single thread to wait on all tasks would be okay, but even that is iffy IMO

Comment (kingcrimsontianyu) [reply cpp/src/io/parquet/reader_impl_preprocess.cu:281]: Thanks! Now that we know the subtasks are submitted eagerly, it is of a trivial issue whether the vector of tasks here are run sequentially or in parallel. So I'll revert the change here.

Comment (kingcrimsontianyu) [reply cpp/src/io/parquet/reader_impl_preprocess.cu]: Completed

**Review by ttnghia — APPROVED** [review]

**Review by vuule — APPROVED** [review]

**Pull request merged**